lib/opFns.js

```diff
@@ -919,20 +936,55 @@ function memLoad (runState, offset, length) {
  * Stores bytes to memory. If an error occurs a string is instead returned.
  * The function also subtracts the amount of gas need for memory expansion.
  * @method memStore
- * @param {Number} offset where to start reading from
- * @param {Number} length how far to read
+ * @param {BN} offset where to start reading from
+ * @param {Buffer} val
+ * @param {BN} valOffset
+ * @param {BN} length how far to read
+ * @param {Boolean} skipSubMem
  * @return {Buffer|String}
  */
 function memStore (runState, offset, val, valOffset, length, skipSubMem) {
   if (skipSubMem !== false) {
     subMemUsage(runState, offset, length)
   }
 
-  var valLength = Math.min(val.length, length)
+  if (length.isZero()) {
+    return
+  }
+
+  // NOTE: in theory this could overflow, but unlikely due to OOG above
+  offset = offset.toNumber()
+  length = length.toNumber()
+
+  if (valOffset.gtn(val.length)) {
+    runState.memory = []
```

Comment:
This doesn't seem to be correct?

Reply:
I found two tests randomStatetest116 randomStatetest237 which expected this exact behavior in order to pass.  Previously ethereumjs was broken under this condition but passed the test because an error caused all of memory to be filled with 'undefined' (i.e. zeroed).

However, this PR caused aforementioned issue to disappear and these two tests to fail.

Reply:
I think the ideal behavior in this case would be to abort the `memStore` and do nothing to memory at all.

Reply:
Ah so actually it wasn't what I thought it was. What is happening is that CALLDATACOPY is storing to memory location zero, but the offset of the call data being stored is greater than the length of the call data. So it's filling the memory with zeros. That makes more sense.

This code does need to change as a result.  I need to brainstorm a test case to better cover this.

Reply:
@jwasinger I don think this code makes a lot of sense, tests still pass and as you said it is the wrong workaround. Do you have a new suggested code?

Reply:
Yeah.  I will amend it after I build some test cases which fail because of this.

Reply:
I created a test case which fails for this commit.  Will merge this PR now